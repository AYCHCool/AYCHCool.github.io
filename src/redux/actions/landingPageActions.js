/* eslint-disable import/prefer-default-export */
import axios from 'axios';
import Toastr from 'toastr';
import {
  SET_TAGS, LOADING, SET_ARTICLES, SET_TRENDING_ARTICLES,
} from './types/landingPage';

const setTags = payload => ({
  type: SET_TAGS,
  payload,
});
const setArticles = payload => ({
  type: SET_ARTICLES,
  payload,
});
const setTrendingArticles = payload => ({
  type: SET_TRENDING_ARTICLES,
  payload,
});

export const getTags = () => async (dispatch) => {
  dispatch({ type: LOADING, payload: true });
  try {
    const response = await axios.get('https://mazus-ah-staging.herokuapp.com/api/v1/articles/tags');
    const { tags } = response.data;
    dispatch(setTags(tags));
    return tags;
  } catch (error) {
    const { data: { errors } } = error.response;
    const message = Object.values(errors)[0];
    Toastr.error(message);
    return true;
  }
};

export const getArticlesByCategory = tags => async (dispatch) => {
  try {
    tags.map(async (tag) => {
      const url = `http://mazus-ah-staging.herokuapp.com/api/v1/articles?tag=${tag}`;
      const res = await axios.get(url);
      dispatch(setArticles(res));
    });
  } catch (error) {
    const { data: { errors } } = error.response;
    const message = Object.values(errors)[0];
    Toastr.error(message);
  }
};

export const getTrendingArticles = () => async (dispatch) => {
  try {
    const response = await axios.get('http://mazus-ah-staging.herokuapp.com/api/v1/articles/trends');
    const { trends } = response.data;
    dispatch(setTrendingArticles(trends));
  } catch (error) {
    const { data: { errors } } = error.response;
    const message = Object.values(errors)[0];
    Toastr.error(message);
  }
};
