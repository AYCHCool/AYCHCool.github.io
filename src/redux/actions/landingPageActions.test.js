import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';
import { getTags, getTrendingArticles, getArticlesByCategory } from './landingPageActions';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const urlToReach = 'http://mazus-ah-staging.herokuapp.com/api/v1';

describe('Test fetch action', () => {
  let store;
  beforeEach(() => {
    store = mockStore({});
  });
  afterEach(() => {
    nock.cleanAll();
  });
  it('should fetch all tags successfully', () => {
    const tags = {
      tags: [
        'fashion',
        'design',
      ],
    };
    nock(urlToReach)
      .persist()
      .get('/articles/tags')
      .reply(200, tags);
    return store.dispatch(getTags())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
  it('should successfully fetch articles by a category', () => {
    const articlesByCategory = {
      data: {
        articles: {
          allArticles: [
            {
              title: 'Three Bad Guys: The Return of Sam',
              description: 'Once again, the three bad guys embarked on a very long Journey',
            },
            {
              title: 'How to catch all your JS Bugs using Baygon360',
              description: 'Both in the virtual and real world, bugs will always...',
            },
          ],
        },
      },
    };
    nock(urlToReach)
      .persist()
      .get('/articles?tag=food&limit=2')
      .reply(200, articlesByCategory);
    return store.dispatch(getArticlesByCategory())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
  it('should fetch all trending articles successfully', () => {
    const trendingArticles = {
      articles: [
        {
          title: 'Three Bad Guys: The Return of Sam',
          description: 'Once again, the three bad guys embarked on a very long Journey',
        },
        {
          title: 'How to catch all your JS Bugs using Baygon360',
          description: 'Both in the virtual and real world, bugs will always...',
        },
      ],
    };
    nock(urlToReach)
      .persist()
      .get('/articles/trends')
      .reply(200, trendingArticles);
    return store.dispatch(getTrendingArticles())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
});
