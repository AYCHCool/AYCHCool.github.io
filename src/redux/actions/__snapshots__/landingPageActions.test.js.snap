// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test fetch action should fetch all tags successfully 1`] = `
Array [
  Object {
    "payload": true,
    "type": "LOADING",
  },
  Object {
    "payload": Array [
      "Fashion",
      "Design",
      "Business",
      "Art",
      "Marketing",
      "Leadership",
      "Startups",
      "Work",
      "Books",
      "Beauty",
      "Movies",
      "Food",
      "Company",
      "Films",
      "Yummies",
      "Leaders",
      "Market",
      "Job",
      "Library",
      "Hustle101",
    ],
    "type": "SETTING_TAGS",
  },
]
`;

exports[`Test fetch action should fetch all trending articles successfully 1`] = `
Array [
  Object {
    "payload": Object {
      "articleCount": 10,
      "articles": Array [
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "How a vegan lifestyle",
          "createdAt": "2019-08-29T16:04:58.809Z",
          "description": "Vegans beauty tips",
          "dislikes": 0,
          "id": "aacd607a-4372-4595-af1a-b5aab1b40466",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "vegans-and-clear-skin-1567094698809-79498",
          "status": "published",
          "tagsList": Array [
            "Fashion",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567086471/Spicy-Vegan-Roasted-Vegetable-Quinoa-Salad-from-HeatherChristo.com__xzonxt.jpg",
          "title": "Vegans and clear Skin",
          "updatedAt": "2019-08-29T16:04:58.809Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "How a vegan lifestyle",
          "createdAt": "2019-08-29T16:04:58.809Z",
          "description": "Vegans beauty tips",
          "dislikes": 0,
          "id": "aacd607a-4372-4595-af1a-b5aab1b40463",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "vegans-and-clear-skin-1567094698809-79498",
          "status": "published",
          "tagsList": Array [
            "Fashion",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567086471/Spicy-Vegan-Roasted-Vegetable-Quinoa-Salad-from-HeatherChristo.com__xzonxt.jpg",
          "title": "Vegans and clear Skin",
          "updatedAt": "2019-08-29T16:04:58.809Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "How a vegan lifestyle",
          "createdAt": "2019-08-29T16:04:58.809Z",
          "description": "Vegans beauty tips",
          "dislikes": 0,
          "id": "aacd607a-4372-4595-af1a-b5aab1b40460",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "vegans-and-clear-skin-1567094698809-79498",
          "status": "published",
          "tagsList": Array [
            "Beauty",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567086471/Spicy-Vegan-Roasted-Vegetable-Quinoa-Salad-from-HeatherChristo.com__xzonxt.jpg",
          "title": "Vegans and clear Skin",
          "updatedAt": "2019-08-29T16:04:58.809Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "Of all my kitchen adventures..",
          "createdAt": "2019-08-29T16:03:06.085Z",
          "description": "How to cook Basmati rice",
          "dislikes": 0,
          "id": "3db114c6-af4f-4f39-b66d-a9a762b312c3",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "the-best-recipe-for-basmati-rice-1567094586085-59808",
          "status": "published",
          "tagsList": Array [
            "Food",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567086777/Basmati_Rice_Salad_sstaln.jpg",
          "title": "The Best recipe for Basmati Rice",
          "updatedAt": "2019-08-29T16:03:06.085Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "Of all my kitchen adventures..",
          "createdAt": "2019-08-29T16:03:06.085Z",
          "description": "How to cook Basmati rice",
          "dislikes": 0,
          "id": "3db114c6-af4f-4f39-b66d-a9a762b312c9",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "the-best-recipe-for-basmati-rice-1567094586085-59808",
          "status": "published",
          "tagsList": Array [
            "Yummies",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567086777/Basmati_Rice_Salad_sstaln.jpg",
          "title": "The Best recipe for Basmati Rice",
          "updatedAt": "2019-08-29T16:03:06.085Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "Once you’ve passed one year being vegan, there’s no going back.",
          "createdAt": "2019-08-29T16:01:17.307Z",
          "description": "Vegans note",
          "dislikes": 0,
          "id": "cf5b88f6-d0cc-4bd4-9835-8a7678e69631",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "ten-things-have-learned-one-year-being-vegan-1567094477307-95330",
          "status": "published",
          "tagsList": Array [
            "Yummies",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567086061/vegan_etqynd.jpg",
          "title": "Ten Things Have Learned One Year Being Vegan",
          "updatedAt": "2019-08-29T16:01:17.307Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "Once you’ve passed one year being vegan, there’s no going back.",
          "createdAt": "2019-08-29T16:01:17.307Z",
          "description": "Vegans note",
          "dislikes": 0,
          "id": "cf5b88f6-d0cc-4bd4-9835-8a7678e69630",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "ten-things-have-learned-one-year-being-vegan-1567094477307-95330",
          "status": "published",
          "tagsList": Array [
            "Food",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567086061/vegan_etqynd.jpg",
          "title": "Ten Things Have Learned One Year Being Vegan",
          "updatedAt": "2019-08-29T16:01:17.307Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "As Disney wraps up the long awaited",
          "createdAt": "2019-08-29T15:59:39.329Z",
          "description": "Thanos is coming to town",
          "dislikes": 0,
          "id": "27a73a9d-5ecb-4c89-868e-5ec8aefa8a6d",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "first-look-at-endgame-1567094379328-27562",
          "status": "published",
          "tagsList": Array [
            "Films",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567085478/Avengers-Endgame-Banner_vbm2zh.jpg",
          "title": "First Look At Endgame",
          "updatedAt": "2019-08-29T15:59:39.329Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "As Disney wraps up the long awaited",
          "createdAt": "2019-08-29T15:59:39.329Z",
          "description": "Thanos is coming to town",
          "dislikes": 0,
          "id": "27a73a9d-5ecb-4c89-868e-5ec8aefa8a1d",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "first-look-at-endgame-1567094379328-27562",
          "status": "published",
          "tagsList": Array [
            "Movies",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567085478/Avengers-Endgame-Banner_vbm2zh.jpg",
          "title": "First Look At Endgame",
          "updatedAt": "2019-08-29T15:59:39.329Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
        Object {
          "author": Object {
            "email": "saintyommex@gmail.com",
            "firstName": "John",
            "id": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            "lastName": "Kamali",
            "profile": Object {
              "avatar": "//www.gravatar.com/avatar/5da3f587e1d3364580a0e20d504465c3?s=200&r=pg&d=mm",
              "bio": null,
              "createdAt": "2019-08-29T15:36:31.489Z",
              "id": "c635b35d-7e04-46f2-a1d0-ce4ff0921e51",
              "updatedAt": "2019-08-29T15:36:31.489Z",
              "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
            },
          },
          "body": "Daniel Craig to continue in Bond 25",
          "createdAt": "2019-08-29T15:58:08.274Z",
          "description": "Bond 25",
          "dislikes": 0,
          "id": "6b889709-7fc6-4c0b-9cc7-4a5d9bc991ff",
          "isPaid": false,
          "likes": 0,
          "ratings": 0,
          "readCount": 0,
          "readTime": 0,
          "reports": 0,
          "slug": "bond-25-1567094288274-84648",
          "status": "published",
          "tagsList": Array [
            "Films",
          ],
          "thumbnail": "https://res.cloudinary.com/mazus/image/upload/v1567085329/daniel-craig-bond-spectre-1537447168_piq8kd.jpg",
          "title": "Bond 25",
          "updatedAt": "2019-08-29T15:58:08.274Z",
          "userId": "abe52a1b-c4c8-42dc-8daa-9852721080ed",
        },
      ],
    },
    "type": "SET_TRENDING_ARTICLES",
  },
]
`;
